ui/src/main/java/org/pentaho/di/ui/core/PropsUI.java

```diff
@@ -1285,4 +1291,104 @@ public void setRecentSearches( String recentSearches ) {
+   * @param repoFileName      the file name
+   * @param lastUsedRepoFiles the {@link Map} of known repository name and user name combinations (example: {@code
+   *                          Local Connection:admin}) to a {@link List} of {@link LastUsedFile}s that have been
+   *                          recently opened from within the repository; this is expected to match the
+   *                          {@link #getLastUsedRepoFiles()} of the {@link PropsUI} singleton instance.
+   * @param lastUsedFiles     The {@link List} of all {@link LastUsedFile}s, opened from the repository, as well as the
+   *                          local file system; this is expected to match the {@link #getLastUsedFiles()} of the {@link
+   *                          PropsUI} singleton instance.
+   */
+  private static void removeRecentImpl(
+    final String repoName, final String repoDirectory, final String repoFileName,
+    final Map<String, List<LastUsedFile>> lastUsedRepoFiles, final List<LastUsedFile> lastUsedFiles ) {
+
+    if ( lastUsedRepoFiles != null ) {
+      final Iterator<Map.Entry<String, List<LastUsedFile>>> lastUsedRepoFilesIter = lastUsedRepoFiles.entrySet()
+        .iterator();
+      while ( lastUsedRepoFilesIter.hasNext() ) {
+        final Map.Entry<String, List<LastUsedFile>> entry = lastUsedRepoFilesIter.next();
+        final String currentRepoNameAndUser = entry.getKey();
+        final String currentRepoName = currentRepoNameAndUser.split( ":" )[ 0 ];
+        if ( currentRepoName.equals( repoName ) ) {
+          final List<LastUsedFile> lastUsedRepoFilesPerRepo = entry.getValue();
+          removeRecentImpl( repoName, repoDirectory, repoFileName, lastUsedRepoFilesPerRepo );
+          break;
+        }
+      }
+    }
+
+    // cleanup last used files, which affects the File > Open Recent menu
+    removeRecentImpl( repoName, repoDirectory, repoFileName, lastUsedFiles );
+  }
+
+  /**
+   * Removes the file identified by the {@code repoDirectory} and {@code repoFileName} within the repo {@code repoName}
+   * from the {@code files} collection.
+   *
+   * @param repoName      the name of the repository containing the file
+   * @param repoDirectory the path to the file's directory
+   * @param repoFileName  the file name
+   * @param files         the {@link List} of {@link LastUsedFile}s from which the recent file will be removed
+   */
+  private static void removeRecentImpl(
+    final String repoName, final String repoDirectory, final String repoFileName, final List<LastUsedFile> files ) {
+
+    if ( files != null ) {
+      for ( LastUsedFile lastUsedRepoFile : files ) {
+        if ( lastUsedRepoFile.getDirectory().equals( repoDirectory )
+          && lastUsedRepoFile.getFilename().equals( repoFileName )
+          && lastUsedRepoFile.getRepositoryName().equals( repoName ) ) {
+          files.remove( lastUsedRepoFile );
+          Spoon.getInstance().addMenuLast();
```

Comment:
This is the only line that I'm not sure about. Can we move the updating of the UI to right after we call removeRecent in the various spots. I know it would be some code duplication, it just feels strange to update the UI inside a file that should just be updating the ui properties file.

Reply:
I had the same reservations - I can certainly move this line out